### ISSUE TITLE
So how to to publish the app? there are not documentations to that.

### ISSUE THREAD

**Issue by prince272** [MicrosoftDocs/visualstudio-docs#7135]
[Enter feedback here]

So how to to publish the app?
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: f5beb7f5-5d28-d5e5-92ba-3e04c75e44b5
* Version Independent ID: 867e3a3f-2387-ba46-c911-10822f8c4a81
* Content: [Create an ASP.NET Core app with React - Visual Studio (Windows)](https://docs.microsoft.com/en-us/visualstudio/javascript/tutorial-asp-net-core-with-react?view=vs-2022)
* Content Source: [docs/javascript/tutorial-asp-net-core-with-react.md](https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/docs/javascript/tutorial-asp-net-core-with-react.md)
* Product: **visual-studio-windows**
* Technology: **vs-javascript**
* GitHub Login: @Mikejo5000
* Microsoft Alias: **mikejo**

**Comment by Simon-Gregory-LG** [comment]
The new SPA template design announcement [ASP.NET Core updates in .NET 6 Preview 4 - Improved single-page app (SPA) templates](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-net-6-preview-4/#improved-single-page-app-spa-templates) briefly mentions the intended deployment scenario:

>  When the app is published, the front-end code in the ClientApp folder is built and collected as before into the web root of the host ASP.NET Core app and served as static files.

There is a lot of discussion on the relative pros / cons of the new template on [ASP.NET Core and SPAs in .NET 6 · Issue #27887 · dotnet/aspnetcore · GitHub](https://github.com/dotnet/aspnetcore/issues/27887)

The primary takeaway from the above links are that:
1) There is a more significant separation between the client-side and server-side code (i.e. they are different projects)
   * This means that the publising scenario is slightly more disconnected
2) The application behaves differently depending on whether it's in Development or Production 
   * The current example only caters for the Development scenario

I would imagine that there will be a better solution when MS tackle the issue properly, but for anybody interested in getting this working while we wait for the [above typescript project publish roadmap ticket](https://developercommunity.visualstudio.com/t/How-do-you-publish-the-new-Standalone-Ty/1559572) to be completed, I have managed to get this sample to publish to an Azure AppService.


# Enhancement to Template for Publishing to Production

## TL;DR

* **Part 1** - Add an `<SpaRoot>` property and `<Target>`  section to the Backend .csproj file tailored to deploy the Frontend project to the backend project's `wwwroot` upon **Publish**
* **Part 2** - Add `UseStaticFiles()` and both/either `UseDefaultFiles()` and/or an endpoint `MapFallback...` in the Backend project's `Program.cs` file
* **Part 3** - If [deploying to AppService (while it's all in preview)](https://azure.github.io/AppService/2021/04/15/Running-dotnet-6-Preview-on-App-Service.html) use either `self-contained` deployment or `container` deployment

## Part 1 - Building & Deploying Frontend Project

Given that the intention for Production is that the typescript project needs be built and copied into the web root, a solution would require an MSBuilt target to be added to the Backend project. 

Using the following three links as inspiration:
* [How to Publish an ASP.NET Core & React SPA to IIS](https://www.roundthecode.com/dotnet/asp-net-core-web-hosting/how-to-publish-an-asp-net-core-react-spa-to-iis)
* [A full explanation for the ResolvedFileToPublish XML element?](https://stackoverflow.com/questions/57526319/a-full-explanation-for-the-resolvedfiletopublish-xml-element)
* [MSBuild well-known item metadata](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-well-known-item-metadata?view=vs-2019)

The `ReactProjectASPNETCore-Backend.csproj` file needs to be amended as follows:
* Add a `<SpaRoot>{path_to_frontend_project}</SpaRoot>` property in the main `<PropertyGroup></PropertyGroup>` to allow referencing the Frontend project's relative path
* Add a `<Target Name="PublishRunWebpack" ...>` target to build the Frontend and deploy into a `wwwroot` folder of the Backend project

**`ReactProjectASPNETCore-Backend.csproj` adjustments:**
``` xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    ...
    <SpaRoot>..\ReactProjectASPNETCore-Frontend\</SpaRoot><!-- Add $(SpaRoot) Property to reference the other Frontend Project -->
  </PropertyGroup>

  ...

  <!-- Customise PublishRunWebpack to deploy into wwwroot -->
  <Target Name="PublishRunWebpack" AfterTargets="ComputeFilesToPublish">
    <!-- As part of publishing, ensure the JS resources are freshly built in production mode -->
    <Exec WorkingDirectory="$(SpaRoot)" Command="npm install" />
    <Exec WorkingDirectory="$(SpaRoot)" Command="npm run build" />

    <!-- Include the newly-built files in the publish output -->
    <ItemGroup>
      <DistFiles Include="$(SpaRoot)build\**" />
      <ResolvedFileToPublish Include="@(DistFiles->'%(FullPath)')" Exclude="@(ResolvedFileToPublish)">
        <RelativePath>wwwroot\%(DistFiles.RecursiveDir)%(DistFiles.FileName)%(DistFiles.Extension)</RelativePath>
        <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
      </ResolvedFileToPublish>
    </ItemGroup>
  </Target>
</Project>
```
> Note that `<RelativePath>` has been adjusted from `%(DistFiles.Identity)` to `wwwroot\%(DistFiles.RecursiveDir)%(DistFiles.FileName)%(DistFiles.Extension)` to customise the target path to be the Backend project's `wwwroot` folder

## Part 2 - Adjust Backend to Serve Frontend Assets

The templates by default do not serve static files, so we will need to add `app.UseStaticFiles()` between `app.UseHttpsRedirection()` and `app.UseRouting()`.

I also added `app.UseDefaultFiles()`, which will mean the `index.html` is served at the root address. 

``` csharp
app.UseHttpsRedirection();

// Add to enable Backend to serve Frontend assets
app.UseDefaultFiles();
app.UseStaticFiles();

app.UseRouting();

```

### Frontend Routing Fallback
However this is quite basic and will not support client-side routing, so I [used this trick from a Stack Overflow post](https://stackoverflow.com/a/61275033/2765855) rather than adding any additional Spa Extension libraries:


``` csharp
app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers();

    // Add any Backend managed routes you'd like to handle 404's
    endpoints.Map("WeatherForecast/{**slug}", HandleApiFallback);

    // Fallback to the Frontend to complete Routing
    endpoints.MapFallbackToFile("{**slug}", "index.html");
});

// Local Function To Handle 404
static Task HandleApiFallback(HttpContext context)
{
    context.Response.StatusCode = StatusCodes.Status404NotFound;
    return Task.CompletedTask;
}
```
> Note: `UseDefaultFiles()` probably strictly isn't necessary if you use this trick but I left it in as it'll probably be needed at some point anyway if the Backend project is to serve web content.

## Part 3 - Publishing `ASP.NET Core .NET 6` App to Azure AppService

Possibly because .NET 6 is still in preview it seems that there are [conditions to deploy a project in AppService](https://azure.github.io/AppService/2021/04/15/Running-dotnet-6-Preview-on-App-Service.html) (i.e. you need to either do **self-contained** or **container** deployment).

#### Self-Contained Deployment

You can use the Right-Click Publish with the **self-contained deployment mode** or if using [dotnet publish](https://docs.microsoft.com/en-gb/dotnet/core/tools/dotnet-publish) command:

``` cmd
dotnet publish --self-contained true -r <RID>
```

`<RID>` is defined here: https://docs.microsoft.com/en-gb/dotnet/core/rid-catalog

In either case, you are required to also qualify any referenced `<RID>`'s in the .csproj file in the `<RuntimeIdentifiers>` section of the Backend project's .csproj file:

``` xml
  <!-- self-contained publish settings -->
  <PropertyGroup>
    <RuntimeIdentifiers>win-x86</RuntimeIdentifiers>
    <PublishTrimmed>true</PublishTrimmed>
  </PropertyGroup>
```
> Note: I also added the [PublishTrimmed option](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/trim-self-contained) to reduce the deployment size

**Comment by Mikejo5000** [comment]
@Simon-Gregory-LG Thanks for posting that info. I am asking about this.

**Comment by gcrev93** [comment]
@prince272 @Simon-Gregory-LG Our team is working on an official way to publish this new project type. It should be released in the first quarter of next year.

**Comment by andrewhood** [comment]
Is this expected to be ready soon? First quarter of 2022 is almost over.

**Comment by yokidane** [comment]
Any resolution for this issue yet? We have been waiting for long time to publish our application  build with standalone angular template.

**Comment by Mikejo5000** [comment]
@andrewhood @yokidane The fix did not make it into 17.2, but is planned for Q2.

**Comment by WilliamAntonRohm** [comment]
@Mikejo5000 -- Mike, thank you for clarifying.

**Comment by StevenTCramer** [comment]
@WilliamAntonRohm you closed this so I am wondering what was the resolution?  I don't see a link to docs or to a PR.

Thanks.

**Comment by Simon-Gregory-LG** [comment]
That's not really any different to [this reply](https://github.com/MicrosoftDocs/visualstudio-docs/issues/7135#issuecomment-993698285) though is it? Which for I'm sure various valid reasons didn't happen, so surely these things should be open until they are actually fixed in one of the manners @StevenTCramer is referring to.

**Comment by jimmytricks** [comment]
@Mikejo5000, apologies to bug but is there any update on where this is on the roadmap? It seems such a crucial piece of documentation / feature to help with getting people up and running with getting it deployed/published. I note the [earlier link](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdevelopercommunity.visualstudio.com%2Ft%2FHow-do-you-publish-the-new-Standalone-Ty%2F1559572&data=04%7C01%7Cmikejo%40microsoft.com%7Ca24aeeff4d9345a0ffae08d99a50bee6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637710493875140252%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=QdpAOR6mGZ9obur36UMucmCRJZpQ6a5cNRI%2BVwz4hBc%3D&reserved=0) for the community page, but this relates to angular and typescript. Is this roadmap also taking into account react and non typescript too?

Cheers

**Comment by Mikejo5000** [comment]
@jimmytricks There is ongoing work on the feature. I am hopeful to get some updates published very soon. I will follow up.

**Comment by melgish** [comment]
Any update for 2023?

**Comment by Mikejo5000** [comment]
@melgish Sorry, yes, the publish procedures in the article were updated to match the product fix:
https://learn.microsoft.com/en-us/visualstudio/javascript/tutorial-asp-net-core-with-react?view=vs-2022#publish-the-project
